ssl/record/dtls1_bitmap.c

```diff
@@ -13,76 +13,49 @@
 /* mod 128 saturating subtract of two 64-bit values in big-endian order */
 static int satsub64be(const unsigned char *v1, const unsigned char *v2)
 {
-    int ret, i;
-
-    if (sizeof(long) == 8)
-        do {
-            const union {
-                long one;
-                char little;
-            } is_endian = {
-                1
-            };
-            long l;
-
-            if (is_endian.little)
-                break;
-            /* not reached on little-endians */
-            /*
-             * following test is redundant, because input is always aligned,
-             * but I take no chances...
-             */
-            if (((size_t)v1 | (size_t)v2) & 0x7)
-                break;
-
-            l = *((long *)v1);
-            l -= *((long *)v2);
-            if (l > 128)
-                return 128;
-            else if (l < -128)
-                return -128;
-            else
-                return (int)l;
-        } while (0);
-
-    ret = 0;
-    for (i=0; i<7; i++) {
-        if (v1[i] > v2[i]) {
-            /* v1 is larger... but by how much? */
-            if (v1[i] != v2[i] + 1)
+    int ret;
+
+    if (sizeof(long) == 8) {
+        unsigned long l1, l2;
+
+        n2l8(v1, l1);
+        n2l8(v2, l2);
+
+        ret = l1 - l2;
+
+        /* We do not permit wrap-around */
```

Comment:
As for question about 50ec3951dcce9bb316eb004947034d8aed75a42b. I don't really recall, most likely it was about writing *general-purpose* saturating subtraction, putting burden of correct usage on caller. And it should be noted that given that sequence numbers are 48 bits wide, caller did use it correctly. I mean original code worked correctly in sense that overflowing 48-bit counter would result in saturation. Indeed, you'd compare 0xffffffffffff and 0, which in 64-bit subtraction is always big. Original concern that that overflow would result in small difference instead of saturation, right?